ui/src/AzureUserInput.ts

```diff
@@ -18,23 +18,27 @@ export class AzureUserInput implements IAzureUserInput {
         this._persistence = persistence;
     }
 
-    public async showQuickPick<T extends QuickPickItem>(items: T[] | Thenable<T[]>, options: QuickPickOptions): Promise<T> {
+    public async showQuickPick<T extends QuickPickItem>(items: T[] | Thenable<T[]>, options: QuickPickOptions): Promise<T>;
+    public async showQuickPick<T extends QuickPickItem>(items: T[] | Thenable<T[]>, options: QuickPickOptions & { canPickMany: true }): Promise<T[]> {
         if (options.ignoreFocusOut === undefined) {
             options.ignoreFocusOut = true;
         }
 
         let persistenceKey: string | undefined;
         const unhashedKey: string | undefined = (<IAzureQuickPickOptions>options).id || options.placeHolder;
-        if (unhashedKey) {
+        if (unhashedKey && !options.canPickMany) {
             persistenceKey = `showQuickPick.${randomUtils.getPseudononymousStringHash(unhashedKey)}`;
         }
 
-        const result: T | undefined = await vscode.window.showQuickPick(this.getOrderedItems(items, persistenceKey, (<IAzureQuickPickOptions>options).suppressPersistence), options);
-        if (result === undefined) {
+        const result: T[] | undefined = await vscode.window.showQuickPick(this.getOrderedItems(items, persistenceKey, (<IAzureQuickPickOptions>options).suppressPersistence), options);
+        if (result === undefined || result.length === 0) {
```

Comment:
I feel like an empty array is a valid response. As a simple example, it could be like when you're trying to unsubscribe from emails and rather than giving you a single "unsubscribe" button, they give you a list like this:
[ ] Send weekly newsletter
[ ] Send coupons
[ ] Send notifications

And you have to unselect all of them to stop getting those annoying emails haha

Reply:
I'm not convinced.  We want UserCancelled to happen automatically, otherwise telemetry is messed up.  In all the likely use cases I can think of, not selecting anything would be a cancel. The unsubscribe scenario isn't realistic for us.

Reply:
UserCancelled will still happen automatically if they hit "escape" correct? Won't that return undefined? I'm assuming an empty array only happens if they select 'enter'. Does VS Code support "validateInput" on quick picks? Can you force the quick pick to stay open if they haven't selected anything? If not, I still feel like you should throw an actual error message like "This field is required." rather than a UserCancelledError which has no feedback to the user.

Reply:
Talking offline to Stephen, I agree with him now. Response inline. 


---
> UserCancelled will still happen automatically if they hit "escape" correct? Won't that return undefined? I'm assuming an empty array only happens if they select 'enter'. 

Yes. 
> Does VS Code support "validateInput" on quick picks? 

I don't see a validateInput (makes sense since the only input is a choice of options already presented to them). 
>Can you force the quick pick to stay open if they haven't selected anything? 

Aside from ignoreFocusOut, I don't see a way currently. Knowing they haven't selected anything seems to happen after the fact.

> If not, I still feel like you should throw an actual error message like "This field is required." rather than a UserCancelledError which has no feedback to the user.

I think this would be more nuanced. In my case a user choosing no option would mean the user is explicit about not wanting to do anything. Hitting "Enter" not undoable anyway. 

In some cases, the user might prefer a headsup, in which case, the caller would decide what to show and the actions on that info message. Example:  

"You have not chosen the registry to which you deploy these containers. What would you like to do? " Option 1: "Create a more suitable registry" Option 2: "Redo this step" Option 3: "Skip this step hereon"

---

I feel having a flag to indicate whether we cancel on empty responses would be ideal. We default to not cancelling.